### ISSUE TITLE
Feature: support folder-note-core

### ISSUE THREAD

**Issue by aidenlx** [pjeby/quick-explorer#2]
https://github.com/aidenlx/folder-note-core

API: https://github.com/aidenlx/folder-note-core/blob/master/src/typings/api.d.ts

**Comment by aidenlx** [comment]
Any feedback?

**Comment by pjeby** [comment]
Hi Aiden.  Please see my previous forum post for what I said my plans were and some of the criteria I had for a "core" plugin.  I specifically listed some of those things because they were issues with the existing folder note plugins, and it doesn't look like the spinoff has really affected them.  There are probably also criteria that I didn't list in the forum post but that would also be relevant.

On a note more directly related to quick-explorer, though, I would mention that the plugin linked above is hardcoding action contexts for the file-menu event, which means its actions do not appear on QE's context menus, making it currently impractical to use with QE, even before considering API integration.

There are several API integration issues as well, some of which I touched on in the forum post, such as an API that doesn't require accessing `app.plugins.plugins["some-plugin"].someAPI` or needing to manually copy typings that should just be exported from the npm package.

Anyway, as I said in the forum post, I will be working on broader folder notes support for QE at some point, but I have criteria both for integrating with a folder notes API, *and* criteria for replacing my existing Note Folder Autorename plugin.  If another plugin is suitable when I get around to "some point" -- some point likely *not* being in September -- then I will consider using its API and as a replacement for my own plugin, otherwise I will fork NFA and make a more flexible version with a reusable/accessible API, i18n support, etc. that can serve as a basis for this and future projects of mine involving folder notes.

So, I cannot guarantee any particular timeframe for feedback or action on integrating your plugin, because I don't yet know when I will have time for the project.  QE+NFA more than suffice for my own folder note use cases at the moment, so it's not a big priority for me.

What I would suggest you do at the moment would be to fix folder note to work with QE by removing the restriction on what file-menu event contexts are supported, and instead of whitelisting which contexts are useful, to instead blocklist specific contexts where they're not useful.  (Not that I'm aware of any place one would *not* want to have them available.)  This would make your plugin useful with other plugins that offer context menus, including Kanban, Recent Files, Pane Relief, and others, in the event that any of them define custom contexts the way QE does.

I will provide other feedback as time/opportunity arises, but I do expect to be very busy with other projects in September.

**Comment by aidenlx** [comment]
Thanks for the reply! Having you checked the new plugin is all I need. 

> such as an API that doesn't require accessing app.plugins.plugins["some-plugin"].someAPI or needing to manually copy typings that should just be exported from the npm package.

For the former, it seems that it's not something I can fix right now since it's related to ob's plugin API design, and for the latter, do you have any thoughts on how to get `declare module obsidian` to works on imported npm packages? It's indeed annoying to copy-paste all event declarations. 

btw, since you're busy, will you accept PRs when I update the API accordingly?

**Comment by pjeby** [comment]
Regarding the typings, they need to be exported.  (Move the obsidian declaration to the api.d.ts file, for example.)

Regarding not using the plugin namespace, just define a unique and versioned name like `folderNoteCoreAPIv1` and set it on `window` in the plugins' `onload()` and delete it during `onunload()`.  In your exported typings, declare something like

```ts
declare global {
    var folderNoteCoreAPI?: WhateverInterfaceTheAPIProvides
}
```

Then, someone using the package can simply use `folderNoteCoreAPIv1` as if it were already imported, though they still need to check if it exists and display a message asking the user to install the plugin (or do some other fallback) if it is not present at layoutReady.

Regarding PRs for integrating your plugin with QE, before I would review such I would need to review your plugin, which is the thing I'm not committing to spend time on.  I hope you don't take offense, but I found your code extremely hard to follow as it is not using the common conventions for organizing Obsidian plugins.  There are other people who do it like you do, but I find theirs very hard to follow as well, because there are levels of indirection that do not appear to serve any purpose and simply make the code harder to navigate instead of easier to read, at least the way I view it.

Also, in the specific case of this plugin, I find many of the names you've chosen difficult to parse and understand at first glance what they are for.  All of this makes me unwilling to commit to a time frame for the review because I do not know how long it will take me to, for example, verify that your renaming logic is free of race conditions for all folder note types -- which is something I'd want to confirm before I would be willing to replace NFA with another plugin.

(Since using your plugin's API in QE would be effectively endorsing it as well as committing to use it myself, I would definitely want to have reviewed it thoroughly before I would even consider PRs to incorporate its API.)

Please note that because my time for this is extremely limited, responding to every new communication from you on this topic -- be it Discord, PM, forum post, and all the other methods by which you've been contacting me -- are *subtracting* from the time available for me to make any progress on it.  In addition, I am beginning to feel stalked and harassed by the sheer volume of communication on this topic you've directed my way in the last few days, even after I've asked you not to PM me, and told you that this was not an immediate priority for me and that I would be getting around to it "some time".  My work on Obsidian is not part of my job, it is something I do as a hobby, a volunteer, and a user, and this amount of pushing is not appreciated.

I do not mind answering questions on how to develop your plugin or what my goals or criteria are for the project.  But please do not expect any actual progress on integration to happen in September -- if it does, it will be because some time happened to open up on my schedule that is not currently booked.  And my replies to your questions may sometimes take a few days, especially in September.

**Comment by aidenlx** [comment]
Terribly sorry for the trouble I’ve brought😣, I wasn’t expecting that you would reply in such details, and it certainly take up too much of your time to organize such detailed reply. I’ve being too excited when seeing someone is working on the same thing as I do, and expectations of collaborating with someone towards the same goal and getting acknowledged has overwhelmed me. Again, terribly, terribly sorry for the inappropriate PM and unintentional pushing! There is definitely something more to do for me to contribute to your projects, especially considering the fact that you’ve been one of the most experienced plugin developers (a big fan of your monkey-patch package, btw, so useful) I’ll stop treating communications in open source community as some sort of online chatting with “already read” mark and expect immediate response (especially that I, as well, won’t always available for plugin development starting from September). And thanks again for the awesome plugin and all the patience and tolerance during the conversation!!! Hoping that I can compensate my inappropriate actions with a PR with higher quality

**Comment by aidenlx** [comment]
> Regarding the typings, they need to be exported. (Move the obsidian declaration to the api.d.ts file, for example.)

I'm aware of the method and tried it before, this time I've solved the problem and finally figured out the whole solution for exposing API, you can check [this post](https://forum.obsidian.md/t/inter-plugin-communication-api-for-plugins/23618) if you are interested.
